### PR TITLE
Extracting ARange

### LINKED ISSUE #15: Expose range as a first class concept
Diet has a Range type that it uses intenerally. We could move this out from inside Diet and expose it as a first class data type as a replacement for Range from the scala standard library

### PR DESCRIPTION
**Pull request by anicolaspp**
fixes #15

### CODE REVIEW

File: core/src/main/scala/dogs/ARange.scala [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+import scala.annotation.tailrec
+
+import scala.Unit
+
+
+/**
+  * Represent a range [x, y] that can be generated by using discrete operations
+  */
+sealed class ARange[A] (val start: A, val end: A) {
```
Comment (stew) [inline core/src/main/scala/dogs/ARange.scala]: can we call this just Range?

File: core/src/main/scala/dogs/ARange.scala [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+  */
+
+package dogs
+
+import dogs.Order._
+import Predef._
+
+import scala.annotation.tailrec
+
+import scala.Unit
```
Comment (stew) [inline core/src/main/scala/dogs/ARange.scala]: this should be redundant with the Predef import

File: core/src/main/scala/dogs/ARange.scala (hunk 1 above)
Comment (anicolaspp) [inline core/src/main/scala/dogs/ARange.scala]: I can :)

File: core/src/main/scala/dogs/ARange.scala (hunk 2 above)
Comment (anicolaspp) [inline core/src/main/scala/dogs/ARange.scala]: I will optimize the the imports.

**Comment by anicolaspp** [comment]
fixes #17

File: tests/src/test/scala/DietSpec.scala [hunk 3]
```diff
@@ -1,14 +1,21 @@
...
 import org.scalatest.{FlatSpec, Matchers}
+import dogs.std.intOrder
+import dogs.Order._
+
 
+// this has something wrong.
+// once, look at the last test in DietTestJoin. The ranges were return by this test class
+// however, they passed the new test. I think the problem is that you are taken
+// more number that the ones you actually need.
+
```
Comment (stew) [inline tests/src/test/scala/DietSpec.scala]: these comments should go, right?

File: tests/src/test/scala/DietSpec.scala [hunk 4]
```diff
@@ -85,21 +93,22 @@ object DietSpec extends Properties("Diet") {
...
+      if(d.contains(i) != r.contains(i)) {
         println(s"for $i, got ${d.contains(i)} expected ${r.contains(i)}")
+      }
     }
 
     (0 to 1000).toList.forall(i => d.contains(i) == r.contains(i))
   }
-
-/* commenting out unti we can merge to Diets
+ // /* commenting out unti we can merge to Diets
```
Comment (stew) [inline tests/src/test/scala/DietSpec.scala]: lets just remove the line

File: tests/src/test/scala/DietSpec.scala [hunk 5]
```diff
@@ -85,21 +93,22 @@ object DietSpec extends Properties("Diet") {
...
   property("merge") = forAll{ (r1: Ranges, r2: Ranges) =>
-    val d = Diet.merge(fromRanges(r1), fromRanges(r2))
+    //val d = Diet.merge(fromRanges(r1), fromRanges(r2))
```
Comment (stew) [inline tests/src/test/scala/DietSpec.scala]: remove zombie line

File: tests/src/test/scala/DietSpec.scala [hunk 6]
```diff
@@ -225,3 +233,104 @@ class DietTest extends FlatSpec with Matchers {
...
+  it should "increase range to the right" in {
+    val diet = Diet[BigInt] + 20 + 22
+    val range = Range[BigInt](21, 30)
+
+    val other = diet.add(range).disjointSets.map(r => r.generate().toScalaList).toScalaList
+
+    other should contain (20.to(30).toList)
+  }
+
+  it should "pass failing property test" in {
```
Comment (stew) [inline tests/src/test/scala/DietSpec.scala]: is this redundant now?

File: tests/src/test/scala/dogs/ARangeTest.scala [hunk 7]
```diff
@@ -0,0 +1,87 @@
+/**
```
Comment (stew) [inline tests/src/test/scala/dogs/ARangeTest.scala]: why is this in its own directory? can we move it to the parent directory with the rest of the tests and name it RangeTest.scala or RangeSpec.scala?

File: tests/src/test/scala/DietSpec.scala (hunk 3 above)
Comment (anicolaspp) [inline tests/src/test/scala/DietSpec.scala]: yes, no need for them anymore

File: tests/src/test/scala/DietSpec.scala (hunk 4 above)
Comment (anicolaspp) [inline tests/src/test/scala/DietSpec.scala]: yes

File: tests/src/test/scala/DietSpec.scala (hunk 6 above)
Comment (anicolaspp) [inline tests/src/test/scala/DietSpec.scala]: yes

File: tests/src/test/scala/dogs/ARangeTest.scala (hunk 7 above)
Comment (anicolaspp) [inline tests/src/test/scala/dogs/ARangeTest.scala]: when I created when to that folder, I will fix it

**Comment by anicolaspp** [comment]
There was a problem when adding a range that extends the node to the right. I removed the two cases (left and right), but I will be adding them again once I have tested them correctly

**Pull request merged**